File: algorithms/150.evaluate-reverse-polish-notation.cpp
class Solution {
public:
    bool isOperator(string s){
        return s == "+" || s == "-" || s == "*" || s == "/";
    }

    int eval(string op, int n1, int n2){
        if (op == "+")      return n1 + n2;
        else if (op == "-") return n1 - n2;
        else if (op == "*") return n1*n2;
        else{
            assert (op == "/");
            return n1 / n2;
        }
    }

    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        st.push(0);
        for (string s : tokens){
            if (isOperator(s)){
                assert(st.size() > 1);
                int n2 = st.top();  st.pop();
                int n1 = st.top();  st.pop();
                st.push(eval(s,n1,n2));
            }
            else
                st.push(stoi(s));
        }
        assert(st.size() > 0);
        return st.top();
    }
};

File: algorithms/150.evaluate-reverse-polish-notation.java
import java.util.Stack;

public class Solution {
	boolean isOperator(String s) {
		return s.equals("+") || s.equals("-") || s.equals("*") || s.equals("/");
	}

	int eval(String op, int n1, int n2) {
		switch (op) {
			case "+":   return n1 + n2;
			case "-":   return n1 - n2;
			case "*":   return n1 * n2;
			case "/":   return n1 / n2; 
			default:	throw new IllegalArgumentException("Invalid operator");
		}
	}

	public int evalRPN(String[] tokens) {
		Stack<Integer> stack = new Stack<>();
		
		for (String s : tokens) {
			if (isOperator(s)) {
				if (stack.size() < 2)
					throw new IllegalArgumentException("Invalid RPN expression");

				int n2 = stack.pop();
				int n1 = stack.pop();
				stack.push(eval(s, n1, n2));
			} else {
				stack.push(Integer.parseInt(s));
			}
		}
		
		if (stack.size() != 1)
			throw new IllegalArgumentException("Invalid RPN expression");

		return stack.pop();
	}
}
